### PR TITLE
parser: refactor parser step2

### PR DESCRIPTION
**Pull request by coocood**
Refactor expression syntax to use MySQL naming and structure.

Partially fix issue #4026

### CODE REVIEW

**Comment by coocood** [comment]
@winkyao @zimulala PTAL

**Comment by coocood** [comment]
@winkyao 
The syntax `SELECT interval 1 DAY + "2017-01-01";` introduces too many `shift/reduce` conflicts, I haven't found a way to avoid it.

File: parser/lexer.go [hunk 1]
```diff
@@ -412,7 +412,7 @@ func startWithAt(s *Scanner) (tok int, pos Pos, lit string) {
 		s.r.incAsLongAs(isIdentChar)
 		tok, lit = doubleAtIdentifier, s.r.data(&pos)
 	} else {
-		tok = at
+		tok = int('@')
```
Comment (shenli) [inline parser/lexer.go]: Why change this?

File: parser/parser_test.go [hunk 2]
```diff
@@ -92,7 +92,7 @@ func (s *testParserSuite) TestSimple(c *C) {
 		"compact", "redundant", "sql_no_cache sql_no_cache", "sql_cache sql_cache", "action", "round",
 		"enable", "disable", "reverse", "space", "privileges", "get_lock", "release_lock", "sleep", "no", "greatest", "least",
 		"binlog", "hex", "unhex", "function", "indexes", "from_unixtime", "processlist", "events", "less", "than", "timediff",
-		"ln", "log", "log2", "log10", "timestampdiff", "pi", "quote", "none", "super", "default", "shared", "exclusive",
```
Comment (shenli) [inline parser/parser_test.go]: This is a bug fix?

File: parser/lexer.go (hunk 1 above)
Comment (coocood) [reply parser/lexer.go]: '@' is a single character token, `at` should be reserved for the keyword `AT`.
Even though we haven't use the keyword `AT` yet.

File: parser/parser_test.go (hunk 2 above)
Comment (coocood) [reply parser/parser_test.go]: Yes, `default` is a reserved keyword.

**Comment by winkyao** [comment]
@coocood Maybe `%expect n` ? :)

**Comment by coocood** [comment]
@winkyao 
That expression looks weird, seems not widely used, let's handle it when actual user need it.

**Comment by shenli** [comment]
LGTM

**Pull request merged**